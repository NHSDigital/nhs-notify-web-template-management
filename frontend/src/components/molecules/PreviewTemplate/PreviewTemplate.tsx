import { Container, Row, Col, Tag } from 'nhsuk-react-components';
import concatClassNames from '@utils/concat-class-names';
import {
  TemplateStatus,
  templateStatustoDisplayMappings,
  templateTypeDisplayMappings,
} from 'nhs-notify-web-template-management-utils/src/enum';
import {
  Template
} from 'nhs-notify-web-template-management-utils/src/types';
import styles from './PreviewTemplate.module.scss';
import { PreviewTemplateProps } from './preview-template.types';

export function PreviewTemplate({
  template,
  preview,
}: Readonly<PreviewTemplateProps>): JSX.Element {
  return (
    <>
      <h1
        data-testid='preview-message__heading'
        className={styles.review__heading}
      >
        {template.name}
      </h1>
      <Container
        className={concatClassNames('nhsuk-u-margin-bottom-6', 'nhsuk-body-m')}
      >
        <div className={styles.preview}>
          <Row className={styles.preview__row}>
            <Col width='one-third' className={styles.preview__col}>
              <div className={styles.preview__col_heading}>Template ID</div>
            </Col>
            <Col width='two-thirds' className={styles.col}>
              {template.id}
            </Col>
          </Row>
          <Row className={styles.preview__row}>
            <Col width='one-third' className={styles.preview__col}>
              <div className={styles.preview__col_heading}>Type</div>
            </Col>
            <Col width='two-thirds' className={styles.col}>
              {templateTypeDisplayMappings(template.templateType)}
            </Col>
          </Row>
          <Row className={styles.preview__row}>
            <Col width='one-third' className={styles.preview__col}>
              <div className={styles.preview__col_heading}>Status</div>
            </Col>
            <Col width='two-thirds' className={styles.col}>
              <Tag
                color={
                  template.templateStatus === TemplateStatus.SUBMITTED
                    ? 'grey'
                    : undefined
                }
              >
                {templateStatustoDisplayMappings(template.templateStatus)}
              </Tag>
            </Col>
          </Row>
        </div>
        <div
          className={concatClassNames('nhsuk-u-margin-top-4', styles.preview)}
        >
          {preview.map(({ heading, value, id }, idx) => (
            <Row key={id} className={styles.preview__row}>
              <Col width='one-third' className={styles.preview__col}>
                <div
                  id={`preview-heading-${id}`}
                  data-testid={`preview__heading-${idx}`}
                  className={styles.preview__col_heading}
                >
                  {heading}
                </div>
              </Col>
              <Col width='two-thirds' className={styles.col}>
                <div
                  id={`preview-content-${id}`}
                  data-testid={`preview__content-${idx}`}
                  className={styles.preview__col_content}
                  dangerouslySetInnerHTML={{ __html: value }}
                />
              </Col>
            </Row>
          ))}
        </div>
      </Container>
    </>
  );
}

PreviewTemplate.Email = ({
  template,
  subject,
  message,
}: {
  template: Template;
  subject: string;
  message: string;
}) => (
  <PreviewTemplate
    template={template}
    preview={[
      { heading: 'Subject', id: 'subject', value: subject },
      { heading: 'Message', id: 'message', value: message },
    ]}
  />
);

PreviewTemplate.Letter = ({
  template,
  heading,
  bodyText,
}: {
  template: Template;
  heading: string;
  bodyText: string;
}) => (
  <PreviewTemplate
    template={template}
    preview={[
      { heading: 'Heading', id: 'heading', value: heading },
      { heading: 'Body text', id: 'body-text', value: bodyText },
    ]}
  />
);

PreviewTemplate.NHSApp = ({
  template,
  message,
}: {
  template: Template;
  message: string;
}) => (
  <PreviewTemplate
    template={template}
    preview={[{ heading: 'Message', id: 'message', value: message }]}
  />
);

PreviewTemplate.Sms = ({
  template,
  message,
}: {
  template: Template;
  message: string;
}) => (
  <PreviewTemplate
    template={template}
    preview={[{ heading: 'Message', id: 'message', value: message }]}
  />
);
