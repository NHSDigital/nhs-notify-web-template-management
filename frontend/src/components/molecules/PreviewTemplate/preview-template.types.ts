import { Template } from 'nhs-notify-web-template-management-utils';

type PreviewTemplateHeadingsType =
  | 'Id'
  | 'Heading'
  | 'Body text'
  | 'Subject'
  | 'Message';

export type PreviewTemplateProps = {
  template: Template;
  templateTypeText: string;
  additionalMetaFields?: {
    title: string;
    id: string;
    content: React.ReactElement;
  }[];
  contentPreview?: {
    heading: PreviewTemplateHeadingsType;
    id: string;
    value: string;
  }[];
};
