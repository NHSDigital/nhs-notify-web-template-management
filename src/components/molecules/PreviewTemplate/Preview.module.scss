/*
* NHSUK Colours
* https://service-manual.nhs.uk/design-system/styles/colour
*/
$white: #ffffff;
$nhsuk-border-color: #d8dde0;
$nhsuk-secondary-button-color: #4c6272;

.preview {
  background-color: $white;
  border: 1px solid $nhsuk-border-color;

  :hover {
    background-color: $white;
  }

  tr {
    border-bottom: 1px solid $nhsuk-border-color;

    td {
      padding: 1rem;
    }
  }

  tr:last-child &__content {
    margin-bottom: 2.4em;
  }

  &__heading {
    color: $nhsuk-secondary-button-color;
    font-weight: 600;
  }
}
