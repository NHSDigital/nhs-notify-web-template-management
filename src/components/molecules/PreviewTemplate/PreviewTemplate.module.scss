/*
* NHSUK Colours
* https://service-manual.nhs.uk/design-system/styles/colour
*/
$white: #ffffff;
$nhsuk-border-color: #d8dde0;
$nhsuk-secondary-button-color: #4c6272;

.preview {
  background-color: $white;
  border: 1px solid $nhsuk-border-color;

  &__row {
    margin: 0;
    padding: 1rem;
    border-bottom: 1px solid $nhsuk-border-color;
  }

  &__col {
    padding: 0;

    &_heading {
      color: $nhsuk-secondary-button-color;
      font-weight: 600;
    }

    &_content {
      word-wrap: anywhere;
    }
  }
}
