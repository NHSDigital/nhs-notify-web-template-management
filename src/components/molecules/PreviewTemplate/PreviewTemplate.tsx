import { Container, Row, Col } from 'nhsuk-react-components';
import concatClassNames from '@utils/concat-class-names';
import styles from './PreviewTemplate.module.scss';
import { PreviewTemplateProps } from './preview-template.types';

export function PreviewTemplate({
  preview,
}: Readonly<PreviewTemplateProps>): JSX.Element {
  return (
    <Container
      className={concatClassNames('nhsuk-u-margin-bottom-6', 'nhsuk-body-m')}
    >
      <div className={styles.preview}>
        {preview.map(({ heading, value, id }, idx) => (
          <Row key={id} className={styles.preview__row}>
            <Col width='one-third' className={styles.preview__col}>
              <div
                data-testid={`preview__heading-${idx}`}
                className={styles.preview__col_heading}
              >
                {heading}
              </div>
            </Col>
            <Col width='two-thirds' className={styles.col}>
              <div
                data-testid={`preview__content-${idx}`}
                className={styles.preview__col_content}
                dangerouslySetInnerHTML={{ __html: value }}
              />
            </Col>
          </Row>
        ))}
      </div>
    </Container>
  );
}

PreviewTemplate.Email = ({
  subject,
  message,
}: {
  subject: string;
  message: string;
}) => (
  <PreviewTemplate
    preview={[
      { heading: 'Subject', id: 'subject', value: subject },
      { heading: 'Message', id: 'message', value: message },
    ]}
  />
);

PreviewTemplate.Letter = ({
  heading,
  bodyText,
}: {
  heading: string;
  bodyText: string;
}) => (
  <PreviewTemplate
    preview={[
      { heading: 'Heading', id: 'heading', value: heading },
      { heading: 'Body text', id: 'body-text', value: bodyText },
    ]}
  />
);

PreviewTemplate.NHSApp = ({ message }: { message: string }) => (
  <PreviewTemplate
    preview={[{ heading: 'Message', id: 'message', value: message }]}
  />
);

PreviewTemplate.Sms = ({ message }: { message: string }) => (
  <PreviewTemplate
    preview={[{ heading: 'Message', id: 'message', value: message }]}
  />
);
