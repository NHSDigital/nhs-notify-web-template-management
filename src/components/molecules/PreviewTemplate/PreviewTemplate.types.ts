type PreviewTemplateHeadingsType =
  | 'Heading'
  | 'Body text'
  | 'Subject'
  | 'Message';

export type PreviewTemplateProps = {
  preview: {
    heading: PreviewTemplateHeadingsType;
    value: string;
  }[];
};
