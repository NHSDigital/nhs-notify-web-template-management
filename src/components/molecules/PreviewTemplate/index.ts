export * from './PreviewTemplate';
export * from './preview-template.types';
