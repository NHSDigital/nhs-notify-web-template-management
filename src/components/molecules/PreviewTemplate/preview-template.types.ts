type PreviewTemplateHeadingsType =
  | 'Heading'
  | 'Body text'
  | 'Subject'
  | 'Message';

export type PreviewTemplateProps = {
  preview: {
    heading: PreviewTemplateHeadingsType;
    id: string;
    value: string;
  }[];
};
