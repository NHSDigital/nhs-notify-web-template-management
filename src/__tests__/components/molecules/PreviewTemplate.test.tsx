import { render, screen } from '@testing-library/react';
import { PreviewTemplate } from '@molecules/PreviewTemplate';

describe('PreviewTemplate component', () => {
  it('matches snapshot', () => {
    const container = render(
      <PreviewTemplate
        preview={[
          {
            heading: 'Heading',
            id: 'heading',
            value: 'Test value-1',
          },
        ]}
      />
    );

    expect(container.asFragment()).toMatchSnapshot();
  });

  it('renders component correctly', () => {
    render(
      <PreviewTemplate
        preview={[
          {
            heading: 'Subject',
            id: 'subject',
            value: 'Subject value',
          },
          {
            heading: 'Heading',
            id: 'heading',
            value: 'Heading value',
          },
          {
            heading: 'Body text',
            id: 'body-text',
            value: 'Body text value',
          },
          {
            heading: 'Message',
            id: 'message',
            value: 'Message value',
          },
        ]}
      />
    );
    expect(screen.getByTestId('preview__heading-0')).toHaveTextContent(
      'Subject'
    );
    expect(screen.getByTestId('preview__content-0')).toHaveTextContent(
      'Subject value'
    );

    expect(screen.getByTestId('preview__heading-1')).toHaveTextContent(
      'Heading'
    );
    expect(screen.getByTestId('preview__content-1')).toHaveTextContent(
      'Heading value'
    );

    expect(screen.getByTestId('preview__heading-2')).toHaveTextContent(
      'Body text'
    );
    expect(screen.getByTestId('preview__content-2')).toHaveTextContent(
      'Body text value'
    );

    expect(screen.getByTestId('preview__heading-3')).toHaveTextContent(
      'Message'
    );
    expect(screen.getByTestId('preview__content-3')).toHaveTextContent(
      'Message value'
    );
  });
});
