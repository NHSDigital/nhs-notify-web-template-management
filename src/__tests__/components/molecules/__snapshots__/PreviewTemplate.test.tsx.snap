// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PreviewTemplate component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="nhsuk-width-container nhsuk-u-margin-bottom-6 nhsuk-body-m"
  >
    <div
      class="preview"
    >
      <div
        class="nhsuk-grid-row preview__row"
      >
        <div
          class="nhsuk-grid-column-one-third preview__col"
        >
          <div
            class="preview__col_heading"
            data-testid="preview__heading-0"
          >
            Heading
          </div>
        </div>
        <div
          class="nhsuk-grid-column-two-thirds col"
        >
          <div
            class="preview__col_content"
            data-testid="preview__content-0"
          >
            Test value-1
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
